test/filters/test_experimental_voxel_grid.cpp

```diff
@@ -0,0 +1,383 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Point CLoud Library (PCL) - www.pointclouds.org
+ * Copyright (c) 2021-, Open Perception
+ *
+ * All rights reserved
+ */
+
+#include <pcl/filters/experimental/voxel_grid.h>
+#include <pcl/filters/voxel_grid.h>
+#include <pcl/io/pcd_io.h>
+#include <pcl/test/gtest.h>
+#include <pcl/point_types.h>
+
+#include <cmath>
+#include <limits>
+
+using namespace pcl;
+using namespace pcl::io;
+
+PointCloud<PointXYZ>::Ptr cloud(new PointCloud<PointXYZ>);
+PointCloud<PointXYZRGB>::Ptr cloud_rgb(new PointCloud<PointXYZRGB>);
+PointCloud<PointXYZRGBA>::Ptr cloud_rgba(new PointCloud<PointXYZRGBA>);
+
+const float PRECISION = Eigen::NumTraits<float>::dummy_precision() * 2;
+
+template <typename PointT>
+void
+EXPECT_POINT_EQ(const PointT& pt1, const PointT& pt2)
+{
+  const auto& pt1_vec = pt1.getVector4fMap();
+  const auto& pt_vec = pt2.getVector4fMap();
+  EXPECT_TRUE(pt1_vec.isApprox(pt_vec, PRECISION))
+      << "Point1: " << pt1_vec.transpose() << "\nPoint2: " << pt_vec.transpose()
+      << "\nnorm diff: " << (pt1_vec - pt_vec).norm() << std::endl;
+}
+
+template <typename>
+void
+EXPECT_POINT_EQ(const PointXYZRGB& pt1, const PointXYZRGB& pt2)
+{
+  const auto& pt1_vec = pt1.getVector4fMap();
+  const auto& pt_vec = pt2.getVector4fMap();
+  EXPECT_TRUE(pt1_vec.isApprox(pt_vec, PRECISION))
+      << "Point1: " << pt1_vec.transpose() << "\nPoint2: " << pt_vec.transpose()
+      << "\nnorm diff: " << (pt1_vec - pt_vec).norm() << std::endl;
+  EXPECT_EQ(pt1.r, pt2.r);
+  EXPECT_EQ(pt1.g, pt2.g);
+  EXPECT_EQ(pt1.b, pt2.b);
+}
+
+template <typename>
+void
+EXPECT_POINT_EQ(const PointXYZRGBA& pt1, const PointXYZRGBA& pt2)
+{
+  const auto& pt1_vec = pt1.getVector4fMap();
+  const auto& pt_vec = pt2.getVector4fMap();
+  EXPECT_TRUE(pt1_vec.isApprox(pt_vec, PRECISION))
+      << "Point1: " << pt1_vec.transpose() << "\nPoint2: " << pt_vec.transpose()
+      << "\nnorm diff: " << (pt1_vec - pt_vec).norm() << std::endl;
+  EXPECT_EQ(pt1.r, pt2.r);
+  EXPECT_EQ(pt1.g, pt2.g);
+  EXPECT_EQ(pt1.b, pt2.b);
+  EXPECT_EQ(pt1.a, pt2.a);
+}
+
+template <typename PointT>
+void
+EXPECT_POINTS_EQ(PointCloud<PointT> pc1, PointCloud<PointT> pc2)
+{
+  ASSERT_EQ(pc1.size(), pc2.size());
+
+  auto pt_cmp = [](const PointT& p1, const PointT& p2) -> bool {
+    return p1.x > p2.x || (p1.x == p2.x && p1.y > p2.y) ||
+           (p1.x == p2.x && p1.y == p2.y && p1.z > p2.z);
+  };
+  std::sort(pc1.begin(), pc1.end(), pt_cmp);
+  std::sort(pc2.begin(), pc2.end(), pt_cmp);
+
+  for (size_t i = 0; i < pc1.size(); ++i)
+    EXPECT_POINT_EQ<PointT>(pc1.at(i), pc2.at(i));
+}
+
+TEST(SetUp, ExperimentalVoxelGridEquivalency)
+{
+  // PointXYZ
+  {
+    PointCloud<PointXYZ> new_out_cloud, old_out_cloud;
```

Comment:
Thoughts on templating test on point type?

Reply:
I will only keep `PointXYZ` in this test for `setUp` because point type is not relevant

But I will try templating the following tests:
https://github.com/tin1254/pcl/blob/8c893c030d94200803774e820fe8e3caedc06072/test/filters/test_experimental_voxel_grid.cpp#L247
https://github.com/tin1254/pcl/blob/8c893c030d94200803774e820fe8e3caedc06072/test/filters/test_experimental_voxel_grid.cpp#L281
https://github.com/tin1254/pcl/blob/8c893c030d94200803774e820fe8e3caedc06072/test/filters/test_experimental_voxel_grid.cpp#L315